### ISSUE TITLE
Evaluation with Multiple GPUs [FPS is low while evaluating video]

### ISSUE THREAD

**Issue by anustupdas** [dbolya/yolact#17]
Hi I am trying to run eval.py and I am getting an average FPS of 8.54 [approx]. I want the FPS to increase. So is there anyway by which the eval.py can use multiple GPUs?

Thanks.

**Comment by dbolya** [comment]
I don;t support running the same image across multiple GPUs, and I'm not sure how that would work. 8.5 FPS seems low. What GPU are you using and are you using --benchmark mode?

**Comment by anustupdas** [comment]
No I did not use benchmark mode. I am getting the details and then post you. Thanks for the reply.

**Comment by dbolya** [comment]
If you run eval.py in non-benchmark mode then the timing includes calculating mAP and loading data. Are you saying you want to speed up computing mAP, or just the model itself? Because to time just the model itself, you have to use --benchmark. One way you could speed up computing mAP is by ensuring that COCO is on an SSD.

**Comment by anustupdas** [comment]
I am using this command to evaluate a video file.

### **python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video=1.avi --video_multiframe=12 --benchmark**
 But it seems to give an average FPS of 10.34 [approx]

Increasing the  **### --video_multiframe** gives it a boost improving the fps from 8 to 10.
But when I am trying to set the  **### --video_multiframe** value more than 14 it runs out of memory as it is using only one GPU.  Giving the error

 "**_RuntimeError: CUDA out of memory. Tried to allocate 669.62 MiB (GPU 0; 5.94 GiB total capacity; 4.75 GiB already allocated; 97.00 MiB free; 286.31 MiB cached)_**"

So is there any way I can make use of more than one GPU while evaluating? or any ther workaround that will increase the average FPS?

Thanks! :)

**Comment by anustupdas** [comment]
![image](https://user-images.githubusercontent.com/33954271/56803164-95628180-683f-11e9-97bb-288c37c4558a.png)

**Comment by anustupdas** [comment]
Any input about this!?? :(

**Comment by dbolya** [comment]
Sorry for the late reply. Video mode isn't completely optimized yet, since I need to make drawing the frames faster than it currently is. I also have no plans at the moment to support evaluation on multiple GPUs, since I don't have multiple GPUs to test on. However, I am working on making the single GPU video evaluation speed much faster (close to exactly the same as raw --benchmark mode).

**Comment by anustupdas** [comment]
Okay. Thanks for the reply. Is there anyway I can modify the code so that it can support multiple GPUs as of now!??
Or I will just wait for your modifications!

**Comment by dbolya** [comment]
I think it would be pretty easy to wrap the network in `evalvideo` in `eval.py` with a `DataParallel` object, but that function is currently bottlenecked by how fast we can draw the masks onto the image and copy the image to the cpu, not how fast the network can be evaluated. You can still try it if you want, but it probably won't make it much faster until I fix that bottleneck.

**Comment by honghande** [comment]
你好，我在使用图片进行实验时，    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
我没有cuda

**Comment by dbolya** [comment]
@honghande, please make a new issue for an unrelated topic. YOLACT does not support CPU only evaluation, so you're going to need to get CUDA and Pytorch with CUDA support.

**Comment by honghande** [comment]
How to configure cuda? Do you have any instructions?

**Comment by anustupdas** [comment]
@dbolya thanks for the suggestion. I will try it.

One quick question shall I implement the DataParallel where the method is being called and the net is being passed as an argument??  evalvideo(net, args.video) [line number 725] or inside the method.

P.S: I am new to this! :)
Thanks

**Comment by dbolya** [comment]
@anustupdas You know what, the change was simple enough and seems to not have broken anything for evaluation on one GPU so I just pushed it. Let me know if it works for multiple GPUs (haven't tested it, should hopefully just work with CUDA_VISIBLE_DEVICES=0,1,2,3,4,5) and video_multiframe=6 or 12.

@honghande This is not the place for pointers on how to install CUDA. That will depend on your hardware and OS. Please look it up on google and _create a new issue_ if you need something. Don't hijack someone else's issue.

**Comment by anustupdas** [comment]
@dbolya thanks a lot. I am trying it out now.

So the command will be something like this??

export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5
python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video=1.avi --benchmark --video_multiframe=6

correct me if I am wrong!

**Comment by dbolya** [comment]
That should be it yeah. Let me know if there are any issues, since DataParallel can do some funky stuff.

**Comment by anustupdas** [comment]
motorai@motorai-System-Product-Name:~/Workstation/yolact$ export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5
motorai@motorai-System-Product-Name:~/Workstation/yolact$ python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video=1.avi --benchmark --video_multiframe=6
Config not specified. Parsed yolact_resnet50_config from the file name.

Loading model... Done.
Initializing model... Traceback (most recent call last):
  File "eval.py", line 938, in <module>
    evaluate(net, dataset)
  File "eval.py", line 726, in evaluate
    evalvideo(net, args.video)
  File "eval.py", line 608, in evalvideo
    eval_network(transform_frame(get_next_frame(vid)))
  File "eval.py", line 592, in transform_frame
    return frames, transform(torch.stack(frames, 0))
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 139, in forward
    inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 150, in scatter
    return scatter_kwargs(inputs, kwargs, device_ids, dim=self.dim)
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/scatter_gather.py", line 35, in scatter_kwargs
    inputs = scatter(inputs, target_gpus, dim) if inputs else []
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/scatter_gather.py", line 28, in scatter
    return scatter_map(inputs)
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/scatter_gather.py", line 15, in scatter_map
    return list(zip(*map(scatter_map, obj)))
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/scatter_gather.py", line 13, in scatter_map
    return Scatter.apply(target_gpus, None, dim, obj)
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/nn/parallel/_functions.py", line 89, in forward
    outputs = comm.scatter(input, target_gpus, chunk_sizes, ctx.dim, streams)
  File "/home/motorai/.local/lib/python3.7/site-packages/torch/cuda/comm.py", line 148, in scatter
    return tuple(torch._C._scatter(tensor, devices, chunk_sizes, dim, streams))
RuntimeError: CUDA out of memory. Tried to allocate 3.62 MiB (GPU 3; 5.94 GiB total capacity; 0 bytes already allocated; 5.47 GiB free; 0 bytes cached) (malloc at /pytorch/aten/src/THC/THCCachingAllocator.cpp:231)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7f095e279021 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7f095e2788ea in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #2: <unknown function> + 0x13f8255 (0x7f09697ff255 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2_gpu.so)
frame #3: <unknown function> + 0x13f900a (0x7f096980000a in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2_gpu.so)
frame #4: at::native::empty_cuda(c10::ArrayRef<long>, at::TensorOptions const&) + 0x36b (0x7f096aabe7fb in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2_gpu.so)
frame #5: at::CUDAFloatType::empty(c10::ArrayRef<long>, at::TensorOptions const&) const + 0x161 (0x7f096971a571 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2_gpu.so)
frame #6: torch::autograd::VariableType::empty(c10::ArrayRef<long>, at::TensorOptions const&) const + 0x179 (0x7f095d53a2b9 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libtorch.so.1)
frame #7: at::TypeDefault::copy(at::Tensor const&, bool, c10::optional<c10::Device>) const + 0x122 (0x7f095efdbd42 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2.so)
frame #8: <unknown function> + 0x977f47 (0x7f095ee01f47 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2.so)
frame #9: at::native::to(at::Tensor const&, at::TensorOptions const&, bool, bool) + 0x295 (0x7f095ee03af5 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2.so)
frame #10: at::TypeDefault::to(at::Tensor const&, at::TensorOptions const&, bool, bool) const + 0x17 (0x7f095efa24f7 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libcaffe2.so)
frame #11: torch::autograd::VariableType::to(at::Tensor const&, at::TensorOptions const&, bool, bool) const + 0x17a (0x7f095d4e2baa in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libtorch.so.1)
frame #12: torch::cuda::scatter(at::Tensor const&, c10::ArrayRef<long>, c10::optional<std::vector<long, std::allocator<long> > > const&, long, c10::optional<std::vector<c10::optional<at::cuda::CUDAStream>, std::allocator<c10::optional<at::cuda::CUDAStream> > > > const&) + 0x391 (0x7f099eaf4b61 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #13: <unknown function> + 0x4ed28f (0x7f099eafa28f in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #14: <unknown function> + 0x11663e (0x7f099e72363e in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #22: THPFunction_apply(_object*, _object*) + 0x5a1 (0x7f099e920961 in /home/motorai/.local/lib/python3.7/site-packages/torch/lib/libtorch_python.so)


@dbolya faced an out of memory error from CUDA
Did I do something wrong!?

**Comment by anustupdas** [comment]
![image](https://user-images.githubusercontent.com/33954271/56957026-82172500-6b63-11e9-9534-bc856c1ded72.png)

**Comment by dbolya** [comment]
That's weird, per the error message itself you shouldn't be out out of memory (it says 5.47 GB free). I'll need to test this myself on a multi GPU setup when I have access to one.

**Comment by anustupdas** [comment]
yes... it is very weird. Maybe you can give some input to fix the bottleneck and see weather it increases the FPS

**Comment by dbolya** [comment]
I still don't know how I'm going to fix the bottleneck, since the issue is not GPU usage, but Python's GIL (i.e., python can't truly multithread).

But to to debug this error, maybe try only two GPUs and video_multiframe=2 to test? The export would then be
```
export CUDA_VISIBLE_DEVICES=0,1
```
or any other two devices (just use `nvidia-smi` to choose 2 that are free, the indices are in the order that `nvidia-smi` lists them).

**Comment by anustupdas** [comment]
I am trying it now. Thank you for the reply.

**Comment by anustupdas** [comment]
@dbolya 

I tried what you suggested.

**export CUDA_VISIBLE_DEVICES=0,1
python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video=1.avi --benchmark --video_multiframe=2**


But encountered an error saying:
**RuntimeError: binary_op(): expected both inputs to be on same device, but input a is on cuda:1 and input b is on cuda:0**

While reading about the problem I found someone suggested to use Torch.no_grad and but sure where to use it in the code. 
https://discuss.pytorch.org/t/torch-no-grad-and-the-updating-of-the-batch-normalization-statistics/25603

**Comment by anustupdas** [comment]
![image](https://user-images.githubusercontent.com/33954271/57067856-c6cec780-6ced-11e9-8c05-947bfd85d0e7.png)

**Comment by dbolya** [comment]
Ah it looks like the multi-GPU support wasn't just plug-and-play. That error means it's trying to do one operation with two matrices on different GPUs. I already use `no_grad` everywhere, so that shouldn't be an issue.

I'll have to debug this on multiple GPUs myself. I'll get back to you as soon as I can do that.

**Comment by anustupdas** [comment]
Okay thanks!!
waiting for your inputs! :)

**Comment by anustupdas** [comment]
if you need I can provide you access to my workstation so that you will have access to multiple GPUs while debugging.

:)

**Comment by dbolya** [comment]
I am still working on this. It's taking so long because I'm also bundling a fix to the bottleneck in the same patch and that's requiring a lot of rewriting.

**Comment by anustupdas** [comment]
> I am still working on this. It's taking so long because I'm also bundling a fix to the bottleneck in the same patch and that's requiring a lot of rewriting.

Thank you so much for the reply. When I am not calling the draw mask function the FPS is improving a bit but not drastically. 

The postprocess function in output.utils is returning the mask contours along with the boxes and classes for each object in a frame, right!?

Correct me If I am wrong!

**Comment by dbolya** [comment]
Yeah running it on a 1060 will do that, since the model itself is going to run slow. But once I fix this multi-gpu thing you'll be zooming with 6 of them.

And yeah that postprocess function returns what you say (except it's not mask contours per say, it's a mask image; i.e., 1 for every pixel where the mask is and 0 elsewhere). What I'm working on right now is making both `prep_display` and `postprocess` support batching, which will allow it to benefit from `video_multiframe` and multiple GPUs.

**Comment by dbolya** [comment]
Ok, finally finished this. Pull the latest commit and first run
```
export CUDA_VISIBLE_DEVICES=0
python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=2 --video=1.avi
```
to run on one GPU to test your new single GPU FPS (since I made a lot of improvements to speed, also `--benchmark` doesn't do anything in `--video` mode).

Then run
```
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5
python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=12 --video=1.avi
```
to test on 6 GPUs. You should also try less than 6 GPUs too because I noticed this can get very CPU bound very quickly (of course modify `--video_multiframe` accordingly). Also let it run for a little bit before reporting the FPS because it goes up over time in my experience.

**Comment by anustupdas** [comment]
Thank you so much for your effort! I will test this and let you know!

**Comment by anustupdas** [comment]
Hello @dbolya,

So I tested the changes that you have made and noted some findings. Just wanting you to know.

I first ran the eval.py with single GPU and the FPS that I am getting is around 12.

 **export CUDA_VISIBLE_DEVICES=0
 python eval_old.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=3 --video=2.avi 
Config not specified. Parsed yolact_resnet50_config from the file name.

Loading model... Done.
Initializing model... Done.

Avg FPS: 14.10**       

But whenever I am increasing the  value of the --video_multiframe parameter the FPS improves a bit but the video displayed gets stuck in between and more the --video_multiframe value is the more it gets stuck.

Next I tried it with 3 GPUs and the performance is not very good.

**export CUDA_VISIBLE_DEVICES=0,1,2
motorai@motorai-System-Product-Name:~/Workstation/yolact$ python eval_old.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=3 --video=2.avi 
Multiple GPUs detected! Turning off JIT.
Config not specified. Parsed yolact_resnet50_config from the file name.

Loading model... Done.
Initializing model... Done.

Avg FPS: 3.44**       



**python eval_old.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=9 --video=2.avi 
Multiple GPUs detected! Turning off JIT.
Config not specified. Parsed yolact_resnet50_config from the file name.

Loading model... Done.
Initializing model... Done.

Avg FPS: 6.52**  

The video displayed got stuck in between even more.

**python eval_old.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=21 --video=2.avi
Multiple GPUs detected! Turning off JIT.
Config not specified. Parsed yolact_resnet50_config from the file name.

Loading model... Done.
Initializing model... Done.

Avg FPS: 9.01** 

So the bottom line is the more the --video_multiframe value the more it is getting stuck.

But with single GPU it is performing quite good now.

**Comment by dbolya** [comment]
The "sticking" is because for instance with video_multiframe=2, two frames are evaluated at once. They then both finish at the same time and I need to decide when to display each frame. If I just display them one after another, it'll be too fast and you'll only "see" one frame. Right now I have it set to use the native FPS of the video as the delay (i.e., if you get 2 frames and the video runs at 30 fps originally, display the first frame then wait 33 ms and display the second). Now this "dumb" approach results in the problem you see when the FPS is too low where between two frames there'll be 33 ms of delay, but between the next two frames there will be the real delay of computation which is much longer. A "smarter" approach would be a lot harder to implement, but it seems like it would be necessary for low FPS evaluation like in your case (and with multiple GPUs).

And hmm, so I guess multiple GPUs still doesn't improve the speed for you huh. Can you check what's your CPU usage when using multiple GPUs? I'll check if I can make anything more efficient but if your CPU is maxed out on the core Python is running on, then there's nothing left I can do in Python to improve your FPS (damn GIL!)

**Comment by dbolya** [comment]
Hey can you try those commands again with the latest commit? I've decoupled processing and playback, and now after a couple of seconds the playback should be smooth no matter the FPS. This should also have resulted in an fps improvement, and now you can use --video_multiframe=whatever you want without worries.

**Comment by anustupdas** [comment]
Surely. I will try it and let you know with the findings.

**Comment by abhigoku10** [comment]
@anustupdas  whats was the results on the latest commit where you able to get more fps with multi-gpus

**Comment by dbolya** [comment]
I'm closing this since it's been open so long. If you have any updates, feel free to reopen.

**Comment by andrenatal** [comment]
I tried these on latest pytorch an cuda 12, on 4 RTX 2080TI and the multi-gpu setup gave me a combination of segmentation faults or `CUDA error: misaligned address"

I managed to run with 1 gpu only though and the best fps (40) on video playback was achieved with `--benchmark` set and --video_multiframe=12`, but only after the frames in buffer increased. Before that was significantly sluggish.
